src/image.py

```diff
@@ -101,16 +93,65 @@ def update_image_data(
                 )
                 if (
                     not unit_auth_data
-                    or (cloud_id := unit_auth_data.get_id()) not in cloud_id_to_image_id
+                    or (cloud_id := unit_auth_data.get_id()) not in cloud_id_to_image_ids
                 ):
                     logger.warning("Cloud auth data not found in relation with %s", unit.name)
                     continue
                 relation.data[self.model.unit].update(
-                    ImageRelationData(
-                        id=cloud_id_to_image_id[cloud_id],
-                        tags=",".join((arch.value, base.value)),
-                    )
+                    _cloud_images_to_relation_data(cloud_images=cloud_id_to_image_ids[cloud_id])
                 )
                 # the relation data update is required only once per relation since every units in
                 # the relation share the same OpenStack cloud credentials.
                 break
+
+
+def _build_cloud_to_images_map(
+    cloud_images: list[list[builder.CloudImage]],
+) -> dict[str, list[builder.CloudImage]]:
+    """Build a map of cloud id to cloud_images.
+
+    Args:
+        cloud_images: The cloud id and image ids to propagate via relation data.
+
+    Returns:
+        The map of cloud_id to cloud images.
+    """
+    cloud_id_to_image_ids: dict[str, list[builder.CloudImage]] = defaultdict(list)
+    for images in cloud_images:
+        for image in images:
+            cloud_id_to_image_ids[image.cloud_id].append(image)
+    return cloud_id_to_image_ids
+
+
+def _cloud_images_to_relation_data(cloud_images: list[builder.CloudImage]) -> ImageRelationData:
+    """Transform cloud images data to relation data.
+
+    Args:
+        cloud_images: The images built for a cloud.
+
+    Returns:
+        The image relation data for a unit.
+    """
+    primary = cloud_images[0]
+    return ImageRelationData(
+        id=primary.image_id,
+        tags=_format_tags(image=primary),
+        images=json.dumps(
+            list(
+                ImageRelationData(id=image.image_id, tags=_format_tags(image=image))
+                for image in cloud_images
+            )
+        ),
+    )
+
+
+def _format_tags(image: builder.CloudImage) -> str:
+    """Generate image tags.
+
+    Args:
+        image: The cloud image.
+
+    Returns:
+        The CSV formatted tags.
+    """
+    return ",".join((image.arch.value, image.base.value))
```

Comment:
Is a method containing a one liner really necessary ?

Reply:
I've put this in place because the tags will be more complex later on when we introduce juju, microk8s, ..etc :)